torch/csrc/autograd/variable.cpp

```diff
@@ -127,22 +127,24 @@ std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
   return _grad_fn;
 }
 
-void VariableViewImpl::rebase_grad_fn(std::shared_ptr<Function> grad_fn) {
+void VariableViewImpl::rebase_history(VarFlags flags, int output_nr, std::shared_ptr<Function> grad_fn) {
   TORCH_ASSERT(output_nr == 0);
-  if (!grad_fn) {
-    TORCH_ASSERTM(!requires_grad, "Can't set null grad_fn on view with requires_grad=True");
+  TORCH_ASSERT(flags.requires_grad == bool(grad_fn));
+  if (grad_fn) {
+    TORCH_ASSERTM(grad_fn->num_inputs == 1, "Functions which modify views in-place must return a single Variable");
+  } else {
     TORCH_ASSERTM(!base.requires_grad(), "base.requires_grad does not match view.requires_grad");
-    return;
   }
-
-  TORCH_ASSERTM(requires_grad, "Can't set grad_fn on view with requires_grad=False");
-  if (grad_fn->num_inputs != 1) {
-    throw std::runtime_error("Functions which modify views in-place must return a single Variable");
+  requires_grad = flags.requires_grad;
+  is_volatile = flags.is_volatile;
+  output_nr = output_nr;
+  base.requires_grad() |= flags.requires_grad;
+  base.is_volatile() |= flags.is_volatile;
```

Comment:
This looks as if you could have ended up with base that both requires_grad (e.g. it required before) and is volatile (set now).

Reply:
If the view has requires_grad=False then the base must already have requires_grad=False (see assertion above)